### PR TITLE
Performance optimizations

### PR DESCRIPTION
**Pull request by sglienke**
Several optimizations for better speed and reduced memory consumption (less string allocations).

The overall speed gain is around 30% with the units I tested (several ones from the RTL and FMX)
As for memory consumption its around 10% less varying a bit on the content of the unit.

More memory could be saved by using string interning which on the other hand would reduce speed a little bit (but not significantly - based on the algorithm to detect identical strings - when using a naive TDictionary<string,string> it still is faster than it was before my optimizations) - but that is subject of another pull request I might do in the future.

### CODE REVIEW

File: Source/DelphiAST.Classes.pas [hunk 1]
```diff
@@ -163,29 +160,23 @@   TOperators = class
...
+      Exit(OperatorsInfo[i]);
 end;
 
 class function TOperators.IsOpName(Typ: TSyntaxNodeType): Boolean;
+var
+  i: Integer;
 begin
-  Result := FOps.ContainsKey(Typ);
+  for i := 0 to High(OperatorsInfo) do
+    if OperatorsInfo[i].Typ = Typ then
```
Comment (NickRing) [inline Source/DelphiAST.Classes.pas]: This looks very similar to the code above (which isn't the best, as if you modify one, you need to remember to modify the other).

File: Source/DelphiAST.pas [hunk 2]
```diff
@@ -2068,7 +2177,7 @@ procedure TPasSyntaxTreeBuilder.StructuredType;
 
 procedure TPasSyntaxTreeBuilder.SubrangeType;
 begin
-  FStack.Push(ntType).SetAttribute(anName, sSUBRANGE);
```
Comment (NickRing) [inline Source/DelphiAST.pas]: Nit Picky: Is "sSUBRANGE" still being used? If not, has it been removed?

File: Source/SimpleParser/SimpleParser.Lexer.pas [hunk 3]
```diff
@@ -1792,8 +1770,13 @@ procedure TmwBasePasLex.IntegerProc;
 end;
 
 function TmwBasePasLex.IsDefined(const ADefine: string): Boolean;
+var
+  i: Integer;
 begin
-  Result := FDefines.IndexOf(ADefine) > -1;
+  for i := 0 to High(FDefines) do
+    if FDefines[i] = ADefine then
```
Comment (sglienke) [inline Source/SimpleParser/SimpleParser.Lexer.pas]: That was my first thought as well when I changed this as TStrings.IndexOf is also case sensitive. But it seems that all directive handling is uppercase (look at GetDirectiveParam). Keep in mind that this commit is not about fixing anything but performance. If there are issues in the affected code this commit was not addressing them.

File: Source/DelphiAST.pas (hunk 2 above)
Comment (sglienke) [inline Source/DelphiAST.pas]: Yes, it has. Look at [the diff](https://github.com/RomanYankovsky/DelphiAST/pull/188/files#diff-a64126dddb118b6a9bc68d27e1b0494cL294).

File: Source/DelphiAST.Classes.pas [hunk 4]
```diff
@@ -163,29 +160,23 @@   TOperators = class
...
-
 class function TOperators.GetItem(Typ: TSyntaxNodeType): TOperatorInfo;
+var
+  i: Integer;
 begin
-  Result := FOps[Typ];
+  for i := 0 to High(OperatorsInfo) do
+    if OperatorsInfo[i].Typ = Typ then
+      Exit(OperatorsInfo[i]);
 end;
```
Comment (sglienke) [inline Source/DelphiAST.Classes.pas]: Might raise an exception in that case just like the dictionary did. But that would only happen if there were a new operator introduced to the language as all the existing operators are in the OperatorsInfo array.

File: Source/DelphiAST.pas (hunk 2 above)
Comment (NickRing) [inline Source/DelphiAST.pas]: Missed it - my bad :(

File: Source/DelphiAST.Classes.pas (hunk 1 above)
Comment (NickRing) [inline Source/DelphiAST.Classes.pas]: I was thinking DRY, that's all.

**Pull request merged**